src/libaktualizr/package_manager/packagemanagerfactory.h

```diff
@@ -5,12 +5,33 @@
 #include "package_manager/packagemanagerinterface.h"
 #include "storage/invstorage.h"
 
+using PackageManagerBuilder =
+    std::function<PackageManagerInterface*(const PackageConfig&, const BootloaderConfig&,
+                                           const std::shared_ptr<INvStorage>&, const std::shared_ptr<HttpInterface>&)>;
+
 class PackageManagerFactory {
  public:
+  static bool registerPackageManager(const char* name, PackageManagerBuilder builder);
   static std::shared_ptr<PackageManagerInterface> makePackageManager(const PackageConfig& pconfig,
                                                                      const BootloaderConfig& bconfig,
                                                                      const std::shared_ptr<INvStorage>& storage,
                                                                      const std::shared_ptr<HttpInterface>& http);
 };
 
+// macro to auto-register a package manager
+// note that static library users will have to call `registerPackageManager` manually
+
+#define AUTO_REGISTER_PACKAGE_MANAGER(name, clsname)                                                         \
+  class clsname##_PkgMRegister_ {                                                                            \
+   public:                                                                                                   \
+    clsname##_PkgMRegister_() {                                                                              \
+      PackageManagerFactory::registerPackageManager(                                                         \
+          name, [](PackageConfig pconfig, BootloaderConfig bconfig, std::shared_ptr<INvStorage> storage,     \
+                   std::shared_ptr<HttpInterface> http) {                                                    \
+            return new clsname(std::move(pconfig), std::move(bconfig), std::move(storage), std::move(http)); \
+          });                                                                                                \
+    }                                                                                                        \
+  };                                                                                                         \
+  static clsname##_PkgMRegister_ clsname##_register_
+
```

Comment:
Oh no, not a macro! :)

Do you think this is possible with templating? I can live with the macro, but I want to resist them if possible.

Reply:
Ok, I've tried:

```
template <typename T>
class AutoRegisterPkgm {
 public:
  AutoRegisterPkgm<T>() {
    PackageManagerFactory::registerPackageManager(
        T::regname, [](PackageConfig pconfig, BootloaderConfig bconfig, std::shared_ptr<INvStorage> storage,
                 std::shared_ptr<HttpInterface> http) {
          return new T(std::move(pconfig), std::move(bconfig), std::move(storage), std::move(http));
        });
  }
};
```

But every package manager would have to have this static `::regname` member, though it cannot be present in `PackageManagerInterface`. This is because I can't parametrize the template with string.

So I'd say it is a bit worse.

An alternative would be to switch back to gcc's `__attribute__(constructor)` like in the previous iteration.

Reply:
I think the `__attribute__(constructor)` works with clang also, so while non-standard ... it sort of is.

Reply:
To summarize our discussion: all three options have pros and cons, but the current macro version is probably the best. Let's stick with that. Thanks for researching the options!